### ISSUE TITLE
Prediction running on CPU instead of GPU 

### ISSUE THREAD

**Issue by ansh1012** [h2oai/h2o4gpu#625]
Configuration--->
- **OS platform, distribution and version (e.g. Linux Ubuntu 16.04)**: Linux 4.4.0-109-generic x86_64
x86_64 GNU/Linux
- **Installed from (source or binary)**: Binary(using PIP)
- **Version**: 0.2.0
- **Python version (optional)**: 3.6
- **CUDA/cuDNN version**: CUDA 8.0

Problem---> 

When i am training the model , then it is running fine on GPU (checked through nvidia-smi) and when i am running the prediction then the GPU memory is being allocated properly and the prediction is also running on GPU.

But when I saved the trained model in pickle file.

and performed following steps-->
- Start a new notebook 
- Loaded the saved pickle model in the notebook.
- Loaded the data file on which i have to run my prediction as panda dataframe.
- Run the prediction on the dataframe

Now the prediction is running on CPU and taking all of its cores instead of using GPU .
I did Watch nvidia-smi and no memory was allocated and and no process ran on it 
also did TOP and CPU usage was clearly shown there. 

Why is it like my prediction after loading model not running on GPU?

My model parameters are -->
`XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1.0, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=8, min_child_weight=1, missing=nan, n_estimators=100,
       n_gpus=1, n_jobs=-1, nthread=None, num_parallel_tree=1, num_round=1,
       objective='binary:logistic', predictor='gpu_predictor',
       random_state=123, reg_alpha=0, reg_lambda=1, scale_pos_weight=1,
       seed=None, silent=False, subsample=1.0, tree_method='gpu_hist')`

**Comment by pseudotensor** [comment]
Hi.  This is a known problem with xgboost itself.  See issue: https://github.com/dmlc/xgboost/issues/3342

What you can do is re-assign the booster's values during prediction, like:

```
def restore_booster_params(model, experiment_id):
    booster = model.get_booster()
    attrs = [a for a in dir(model) if
             not a.startswith('__') and not a.startswith('_') and not callable(getattr(model, a))]
    n_gpus = model.kwargs['n_gpus']
    skip_gpu_param = ngpus_vis_global == 0 or ngpus_vis_global != n_gpus  # FIXME - shouldn't have to have exactly same number of GPUs

    for attr in attrs:
        value = getattr(model, attr)
        str_attr = str(attr)
        if str_attr == 'nthread':
            str_attr = 'n_jobs'
        if str_attr == 'seed':
            str_attr = 'random_state'
        if "gpu" in str_attr or isinstance(value, str) and "gpu" in value:
            if skip_gpu_param:
                continue
        if str_attr != 'booster':
            booster.set_param({str_attr: getattr(model, str_attr)})
    for k, v in zip(model.kwargs.keys(), model.kwargs.values()):
        if isinstance(k, str) and "gpu" in k or isinstance(v, str) and "gpu" in v:
            if skip_gpu_param:
                continue
        booster.set_param({k: v})

    if ngpus_vis_global > 0 and n_gpus > 0 and config.use_gpu_for_prediction:
        if config.gpu_locking_type == "global":
            gpu_lock = AcquireGPUs(n_gpu=n_gpus, experiment_id=experiment_id)
            booster.set_param({'gpu_id': gpu_lock.start_id()})
        # otherwise don't override
        booster.set_param({'predictor': 'gpu_predictor'})
    else:
        booster.set_param({'predictor': 'cpu_predictor'})
```

you should set the missing variables for your own system.

**Comment by ansh1012** [comment]
As in the model parameters , predictor is already set to gpu_predictor and I am loading the model in the same node it has been trained on .
What I am able to understand from this code is that it is setting the predictor param to gpu_predictor or cpu_predictor based on the config. of this node.
Maybe I am wrong !!

**Comment by pseudotensor** [comment]
Yes, what I said above is that loading from a pickled state loses the the information in the booster.  dmlc don't properly pickle the state.

**Comment by ansh1012** [comment]
Thanks a lot !
It worked :) 👍

**Comment by pseudotensor** [comment]
Great!
